### PR TITLE
mkasustrx: New tool for creating TRX-images compatible with AsusWRT

### PR DESCRIPTION
**Pull request by remittor**
This tool allows you to create `trx`-images for many Asus routers. Supported v2 and v3 format of `trx`-images.
Currently v3 `trx`-format has been tested on routers: RT-AX59U, TUF-AX4200, TUF-AX6000.

Usage example for v3 `trx` format:  https://github.com/openwrt-xiaomi/openwrt/commit/6f0e79ffffffc26fd3c38f46c6261d7977ff7bb0

Usage example for v2:
```
define Device/asus_rt-ax53u
...
  DEVICE_MODEL := RT-AX53U
...
  IMAGES += factory.trx
  IMAGE/factory.trx := append-kernel | append-ubi | \
	asus-trx -v 2 -n $$(DEVICE_MODEL) -b 386 -e 65535 
endef
TARGET_DEVICES += asus_rt-ax53u
```

### CODE REVIEW

**Comment by rmilecki** [comment]
I don't see how it's related to TRX format at all. It doesn't use TRX magic, or TRX structure (header). It seems to be uImage with ASUS tail appended. It may be the same tail as in `asustrx.c` but it doesn't really make it TRX related.

**Comment by remittor** [comment]
> It may be the same tail as in `asustrx.c` but it doesn't really make it TRX related.

asustrx.c - for TRX v0 images (magic = HDR0)
mkrtn56uimg.c - for TRX v1 images (magic = uImage)
mkasustrx.c - for TRX v2 & v3 images (magic = uImage)

> I don't see how it's related to TRX format at all. It doesn't use TRX magic, or TRX structure (header).

Asus programmers historically call their images this way. You can see this in the logs:
```
  v5 = j_check_imageheader((int)v14, (int)v29);
  if ( !v5 )
  {
    v3 = a1;
    j_cprintf("Check image heaer fail !!!\n");
    goto LABEL_22;
  }
  v3 = a1;
  v29[0] = bswap32(*(_DWORD *)(v8 + 12));
LABEL_17:
  if ( !j_check_trx(a1, v8) )
  {
    v5 = 2;
    j_cprintf("check trx fail!!\n");
    return v5;
  }
```
In the stock bootloader AsusWRT images are also called that way. Do you need disasm lines and from there?

**Comment by rmilecki** [comment]
If you look at original TRX, you can see there was TRX v1 (with up to 3 partitions) and TRX v2 (with up to 4 partitions):
https://openwrt.org/docs/techref/headers

`asustrx.c` uses header with 3 paritions so I'd call it TRX v1.

I can see Asus developers didn't change TRX names in their code while switching from TRX fo uImage. I'm not sure if that's a good example to follow. Personally I think I'd look to name this tool `asusuimage` or similar.

**Comment by remittor** [comment]
> I can see Asus developers didn't change TRX names in their code while switching from TRX fo uImage. I'm not sure if that's a good example to follow.

Well, that's your opinion. I have been familiar with Asus routers for a very long time and am accustomed to naming images the same way as the Asus developers.
Users of Asus routers download files with the `.trx` extension from the official website (in ZIP-containers).

If you want to change the name, then create your own PR, and I will close mine PR.

**Comment by rmilecki** [comment]
I'm still not convinced. Just because vendor developers make a silly decision doesn't mean we should follow it. It happens all the time. Vendors have their wrongly designed/developed drivers, DTS, scripts that need to be refactored/cleaned up before they go upstream. We very often improve poor quality vendors code. I don't see why we shouldn't do it here. It's not like it requires a lot of time.

**Review by rmilecki — CHANGES_REQUESTED** [review]
1. Please rename this tool to avoid using TRX name
2. Improve coding style:
    * Use tabs instead of spaces
    * Use `static` for functions where possible
    * Define variables at functions beginning
    * Move structs defines & global variables to the top of file
    * Fix endianess (don't use `htons` & friendly directly)
3. Fix compilation warnings! There are 10+ of them

**Comment by remittor** [comment]
> Please rename this tool to avoid using TRX name

Denied!

> Use tabs instead of spaces

The requirement is not justified! First change these files:
```
cyg_crc16.c
cyg_crc32.c
dgfirmware.c
dgn3500sum.c
encode_crc.c
lzma2eva.c
imagetag_cmdline.c
mkedimaximg.c
mkrasimage.c
sha1.c
srec2bin.c
zyimage.c
```

> Fix endianess (don't use htons & friendly directly)

The requirement is not justified! It's not a mistake!
Look this: 
```
> grep -m1 'hton.(' *.c

add_header.c:   header.crc = htonl(crc32buf(buf, buflen));
buffalo-tag.c:          tag->region_mask = htonl(region_mask);
edimax_fw_header.c:     hdr->start_addr = htonl(start_addr);
fix-u-media-header.c:   hdr->ih_UMedia_magic = htonl(UM_MAGIC);
imagetag.c:  uint32_t network = htonl(value);
jcgimage.c:     h->jh_magic = htonl(JH_MAGIC);
mkcameofw.c:    hdr->checksum = htonl(csum);
mkchkimg.c:     hdr->header_len = htonl(header_len);
mkdapimg2.c:    imghdr.checksum = htonl(cksum);
mkdapimg.c:             imghdr.checksum = htonl(cksum);
mkfwimage2.c:   header->crc = htonl(crc32(0L, (unsigned char *)header,
mkfwimage.c:    header->crc = htonl(crc32(0L, (uint8_t*) header,
mkmerakifw.c:   value = htonl(value);
mkmerakifw-old.c:               value = htonl(value);
mkrasimage.c:     return htonl(sysv_chksm(data, size));
mkrtn56uimg.c:  hdr->ih_dcrc = htonl(crc32(0, (Bytef *)data, len));
mksenaofw.c:    header->head = htonl(header->head);
mktplinkfw2.c:  hdr->version = htonl(bswap_32(hdr_ver));
mktplinkfw.c:   hdr->version = htonl(hdr_ver);
mkzyxelzldfw.c: fw.header.checksum = htonl(checksum_calculate(fp_dst, kernel_offset));
motorola-bin.c:                 if (htonl(crc32buf(trx + offsetof(struct motorola, flags), len - offsetof(struct motorola, flags))) != firmware->crc)
seama.c:                        if (shdr.magic != htonl(SEAMA_MAGIC)) ERRBREAK("Invalid SEAMA magic. Probably no more SEAMA!\n");
spw303v.c:      crc = htonl(crc32(crc, fake_data, 64));
tplink-safeloader.c:    header->length = htonl(data_len);
uimage_padhdr.c:        imgh->ih_hcrc = htonl(crc_recalc);
uimage_sgehdr.c:        imgh->ih_hcrc = htonl(crc_recalc);
wrt400n.c:      crc = htonl(crc);
zytrx.c:                h->modelid = htonl(p->modelid);
zyxbcm.c:       crc = htonl(crc32(IMAGETAG_CRC_START, buf, 236));
```

> Use static for functions where possible
> Define variables at functions beginning
> Move structs defines & global variables to the top of file
> Fix compilation warnings! There are 10+ of them

These requirements are justified by the general code style of the entire Linux community.

**Comment by rmilecki** [comment]
I'm sorry but I quite dislike your overall attitude. You justify your contribution quality by referencing other code (no matter what quality it is on its own). You yell and don't explain your decisions just denying changes. Finally you insult other developers (like in another pull request):
> Robimarco was too lazy to control the fan speed.

If someone is going to take this, I want my `Nacked-by` to be included. I don't merge this.

**Comment by rmilecki** [comment]
One more note: `Signed-off-by: remittor`. I expect we still require names in `firmware-utils` unless I missed sth / I'm wrong.

**Comment by rmilecki** [comment]
> > Fix endianess (don't use htons & friendly directly)
> 
> It's not a mistake!

Since author seems too busy yelling at me. Using `htons()` is **correct** as it converts host endianess to network endianess (BE). It's exactly what we want here. For some unknown reason I thought it doesn't take care of endianess correctly.

**Comment by ynezz** [comment]
Oleg,  first of all[there is one important rule](https://openwrt.org/rules#openwrtlede_project_rules) `12. Be nice to each other` so would be nice to keep this in mind when contributing to the OpenWrt projects, thanks! :) 

* Understand that feedback is about the code, not about you personally.
* Try to detach your emotions from the feedback and focus on the technical aspects being discussed.
* Approach the review process as a collaborative effort to produce high-quality code together.
* Remember that everyone's goal is to enhance the project collectively.

> 1. Please rename this tool to avoid using TRX name

Yes, naming is hard, usually leads to bike shedding, but I'm leaning more towards Rafał's reasoning here (disclaimer,: I'm biased :)), so I would recommend to follow the advice if you want this ever merged.

> Denied!

Remember, the goal is to improve the project, not to win an argument.

>     2. Improve coding style:
>        
>        * Use tabs instead of spaces
>        * Use `static` for functions where possible
>        * Define variables at functions beginning
>        * Move structs defines & global variables to the top of file

Yes, please.
 
>     3. Fix compilation warnings! There are 10+ of them

Yes, please.

> These requirements are justified by the general code style of the entire Linux community.

And as we can see, it leads to quite usable systems, so I would say, that those engineering principles are sound.

> First change these files:

Trust me, [everyone would like to get all those issues fixed](https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=shortlog;h=refs/heads/wip/firmware-utils-fixes), but its huge amount of work and IMO we first should cover the current codebase with some CI and test cases (thats one of the reasons for this separate project), so once we start tidying up the codebase and do some refactoring, we can be assured, that nothing breaks in the end. I'm simply afraid, that some of the code currently produces correct images as a side effect, not intentionally, so fixing the issues would actually lead to a broken image and the code would need be fixed.

In other words, Rafał's is trying to prevent adding more items to the already very long TODO list...

**Comment by remittor** [comment]
@rmilecki 
> You yell and don't explain your decisions just denying changes.

I didn't scream! It seemed to you! I justified and explained all my decisions.

> Finally you insult other developers (like in another pull request)

I didn't insult anyone! Robimarco simply wrote a fully working driver, but did not try to bring his work to completion. Other people had to do it.

> Since author seems too busy yelling at me.

Nobody yells at you!

**Comment by remittor** [comment]
@ynezz 
> * Understand that feedback is about the code, not about you personally.

I know this very well. I don’t understand what personalities have to do with it at all. I discuss my and other people's code.

> * Try to detach your emotions from the feedback and focus on the technical aspects being discussed.

That's what I do. Where did you see my emotions?
I communicate using facts as arguments.

> * Approach the review process as a collaborative effort to produce high-quality code together.

This is true. But this does not mean that I must unconditionally fulfill any requirement of the maintainers!

**Comment by ynezz** [comment]
>   I communicate using facts as arguments.

Insulting other community members like Robert for example is not either fact nor argument, so against the project rule 12.

>  But this does not mean that I must unconditionally fulfill any requirement of the maintainers!

Sure, I agree, but why did you created the pull request in the first place? If your intention was to get this merged into the project, then you're somehow expected to address review comments, right? Otherwise I don't see a path forward and you can simply close this.

**Comment by remittor** [comment]
@ynezz 
> Insulting other community members like Robert for example...

Yes, no one insulted Robimarco! Where did you see this?

> Sure, I agree, but why did you created the pull request in the first place?

I hoped for the adequacy of the maintainers.

> If your intention was to get this merged into the project, then you're somehow **expected to address review comments,** right?

Am I not answering? I just answer and present the facts with arguments.

> Otherwise I don't see a path forward and you can simply close this.

You have the right to do this.

**Comment by rmilecki** [comment]
> No one insulted Robimarco!

I believe many people find it offensive to call someone lazy. Especially in context where you improve that person's code.

> Nobody yells at you!

You keep using exclamation marks all around. That's how I understood it although I'm not native English.

**Comment by remittor** [comment]
**My conclusion is this:**

The **naming and tabulation** turned out to be more important than simplifying the installation of the firmware for inexperienced users.

It's very annoying...

**Comment by remittor** [comment]
@rmilecki 
> I believe many people find it offensive to call someone lazy.

This is just your personal opinion.

> You keep using exclamation marks all around. That's how I understood it although I'm not native English.

Read this in full: https://en.wikipedia.org/wiki/Exclamation_mark

**Comment by rmilecki** [comment]
Having proper firmware tools is great and important! Clean code just helps generally in maintenance.

If you were asked to spend hours refactoring some global code that wouldn't be fair. Here it's (as  you noticed) just a matter of improved coding style & minor fixups. So further developers can look at your code and work with it easily. How much time do you realistically need to improve that? I don't think it's unfair to ask for some extra effort before merging code that will stay with us for many many years.

**Comment by remittor** [comment]
@rmilecki 
> If you were asked to spend hours refactoring some global code that wouldn't be fair. Here it's (as you noticed) just a matter of improved coding style & minor fixups.

You're making things up again. Where did I write that I refuse to edit the style of writing code?
I just wrote that I refuse to rename and refuse to replace spaces with tabs.
Spaces do not prevent other users from reviewing the code!

**Comment by ynezz** [comment]
>  I just wrote that I refuse to rename and refuse to replace spaces with tabs.

Ok, then lets not waste more time on this.

**Comment by rmilecki** [comment]
> Spaces do not prevent other users from reviewing the code!

So previously you wrote: "These requirements are justified by the general code style of the entire Linux community." Doesn't https://www.kernel.org/doc/html/v6.7/process/coding-style.html say to use tabs?

I used to work with dozens of maintainers of open source projects, trust me. You can keep proving you're all right, it's other people who misunderstood you and it's all about their opinions. That usually doesn't get you anywhere. It's from experience.

At the end of the day it's a matter if you want your code merged or to prove to be right. And I'm talking about my own contributions here mostly.

**Comment by remittor** [comment]
@rmilecki 
> So previously you wrote: "These requirements are justified by the general code style of the entire Linux community." Doesn't https://www.kernel.org/doc/html/v6.7/process/coding-style.html say to use tabs?

If you consider Linux kernel maintainers to be everyone unix like community, then this is not so.
(above I meant Unix like, and not just the Linux kernel)

**Comment by msva** [comment]
Why on the earth there is so much clownery and shit-throwing here out of thin air?

Are project goal is to have/do nice things or to force people to "do as i said or we'll throw your contributions to the trash even if it will harm us"?

Who prevented anyone who wants another naming and tabs from fixing that during the merge of original contribution?


As a consequence of rage-closing this PR, users of newer ASUS routers are forced to download shitty "transit trx files" from google drives of doubtful persons to reflash their routers to openrwt, instead of using "original" ones.

I think that was totally counter-productive

**Comment by Ansuel** [comment]
Did we for real caused all this mess for a simple change of space to tabs that advanced editor can do with 3 clicks? 

Not following what is happening here...

**Comment by Ansuel** [comment]
I don't have context but the prereq for contributions is accept and comply with the reviewers request if it does 1% sense. If we start to fight and waste time instead of actually do the change then we start talking about politics instead of how to make this in a mergeable state.

**Comment by remittor** [comment]
@Ansuel 
> Did we for real caused all this mess for a simple change of **space to tabs** that advanced editor can do with 3 clicks?

But that's not the main reason. The whole dispute started because of the renaming of the utility and internal identifiers.

@msva
> users of newer ASUS routers are forced to download **shitty "transit trx files"** from google drives of **doubtful persons**

Why am I a dubious person?
Why are my TRX images shitty?

Even if @Ansuel and @robimarko generate TRX images, OpenWrt 24 will still not work on Asus routers with the new bootloader (look this: https://forum.openwrt.org/t/205261/47 )
And users are already forced to use my build v24.10, since I fixed the behavior of the `mtd` driver.

**Comment by msva** [comment]
> I don't have context but the prereq for contributions is accept and comply with the reviewers request if it does 1% sense. If we start to fight and waste time instead of actually do the change then we start talking about politics instead of how to make this in a mergeable state.

the point is that in cases when contributer have no motivation to perform requested changes, why don't reviewers make them themselves to make things happen instead of freezing them or throwing them away?

**Comment by remittor** [comment]
@msva 
> why don't reviewers make them themselves to make things happen instead of freezing them or throwing them away?

@Ansuel , at @robimarko's request, made the changes himself and added my utility to the repository.
Without this utility it was simply impossible to make a working firmware for RT-AX89X

**Comment by msva** [comment]
> Why am I a doubtful person?
> Why are my TRX images shitty?

don't take that on your account
it was about situation in general, where we have tons of guides on different forums (like xda/4pda/whatever) where any random johnny makes his own guide and supply his own builds (which can potentially be harmful).
That often happens when there are no official things available, and such johnnies can use that situation to even just take existing semi-official instructions and just replace links to his crap

**Comment by Ansuel** [comment]
> the point is that in cases when contributer have no motivation to perform requested changes, why don't reviewers make them themselves to make things happen instead of freezing them or throwing them away?

cause everyone is busy and this is not paid work. If the feature gets enough relevance other contributors can step in and do the change. It's problematic tho if we start making a fight

**Comment by msva** [comment]
> The whole dispute started because of the renaming of the utility and internal identifiers.

which, by the way, can have no way to affect the work of the utility, so it is either you can agree to change as requested, or, if someone thinks it is mandatory and it can be used as is, I think he is free to change that as he want after merging it in current state.

(just imho)

Anyway, if that is such a principial thing that really such a blocker, and no one can step into it, I can try to rename the things and convert tabs maybe somewhere next week or so.

// Although, actually I personally wouldn't benefit anyhing from it, as I already sold all my ASUSes, but if nobody else want to do this, I can take it

I'd like that things would be as upstreamed as possible

**Comment by robimarko** [comment]
@msva This tool is already part of firmware-utils, @remittor already stated that, its just called `asusuimage`

**Comment by JonSnow88** [comment]
> @msva This tool is already part of firmware-utils, @remittor already stated that, its just called `asusuimage`

Hopefully we'll now see official TRX images.

**Comment by robimarko** [comment]
@JonSnow88 I dont understand, we have had official TRX images for flashing from stock since the device was added
